src/mpid/ch4/include/mpidpre.h

```diff
@@ -254,6 +253,7 @@ typedef struct MPIDI_part_request {
                                          * Valid also in posted_rreq so that single dequeue
                                          * routine can be used. */
     MPI_Datatype datatype;
+    MPIR_Request *peer_req_ptr;
```

Comment:
I am a bit concerned that it appears to be arbitrary to have this field here. I understand there is a code logic, but it is not obvious at top-level without understanding the lower-layer logic. Can you explain your plan of using `peer_req_ptr`?

Reply:
My idea is to allow lower layers to do additional handshake(s) without disturbing the generic matching process. Once the peer req is set the by the generic layer, it can be used elsewhere. For instance in ch4/ucx, I can send additional information about the actual number of pt2pt operations to be used, or for rkey exchange for an RMA-based protocol.

At the moment, there's no reason the lower layer can't just read the peer req out of the `MPIDIG_part_request`, so maybe this change can wait. I am eventually hoping that `MPIDIG_part_request` and the NM/SHM structs can be a union, but the complex initialization (matched, nonblocking, with no explicit completion) makes this somewhat difficult.

Reply:
I think I agree with the idea of making `MPIDIG_part_request'  part of the union. In the current design, `send_epoch, recv_epoch, peer_req_ptr` are all necessary always needed fields, and I think they all should be moved to `MPIDI_part_request_t`.

Reply:
`send_epoch` and `recv_epoch` won't be needed in the native pt2pt implementation, for what it's worth.